src/main/java/com/google/api/generator/engine/ast/ArrayExpr.java

```diff
@@ -0,0 +1,139 @@
+    return ArrayExpr.builder().setExprsList(Arrays.asList(exprs)).build();
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    private static final String EMPTY_EXPRS_MESSAGE = "List of expressions cannot be empty";
+
+    private static final String SAME_TYPE_EXPRS_MESSAGE = "All expressions must have the same type";
+
+    abstract List<Expr> exprs();
+
+    protected abstract ArrayExpr.Builder setType(TypeNode type);
+
+    /**
+     * To add a ValueExpr as to our list. Can be used repeatedly to add multiple parameters.
+     * same-type validation is performed
+     *
+     * @param expr
+     * @return Builder
+     */
+    public ArrayExpr.Builder addExpr(ValueExpr expr) {
+      return addExprToList(expr);
+    }
+
+    /**
+     * To add a VariableExpr as to our list. Can be used repeatedly to add multiple parameters.
+     * same-type validation is performed
+     *
+     * @param expr
+     * @return Builder
+     */
+    public ArrayExpr.Builder addExpr(VariableExpr expr) {
+      return addExprToList(expr);
+    }
+
+    /**
+     * To add a string expression same-type validation is performed
+     *
+     * @param expr
+     * @return Builder
+     */
+    public ArrayExpr.Builder addExpr(String expr) {
+      return addExprToList(ValueExpr.withValue(StringObjectValue.withValue(expr)));
+    }
+
+    /**
+     * To set the list of expressions for the anonymous array Validates that every expression is of
+     * the same type
+     *
+     * @param exprs
+     * @return Builder
+     */
+    private ArrayExpr.Builder setExprsList(List<Expr> exprs) {
+      Preconditions.checkState(exprs.size() > 0, EMPTY_EXPRS_MESSAGE);
+      // validate types
+      TypeNode baseType = exprs.get(0).type();
+      for (int i = 1; i < exprs.size(); i++) {
+        TypeNode currentType = exprs.get(i).type();
+        Preconditions.checkState(currentType.equals(baseType), SAME_TYPE_EXPRS_MESSAGE);
+        baseType = currentType;
+      }
+      setType(TypeNode.arrayOf(baseType));
+      return setExprs(exprs);
+    }
+
+    // this method is private, and called only by addExpr(Expr expr)
+    private ArrayExpr.Builder addExprToList(Expr expr) {
+      List<Expr> exprList = exprs();
+      if (exprList == null) {
+        exprList = new ArrayList<>();
+      }
+      exprList.add(expr);
+      return setExprsList(exprList);
+    }
+
+    // this setter is private, and called only by setExprsList() to ensured sanitized contents
+    abstract ArrayExpr.Builder setExprs(List<Expr> descriptionExprs);
+
+    abstract ArrayExpr autoBuild();
+
+    public ArrayExpr build() {
+      ArrayExpr anonymousArrayExpr = autoBuild();
+      Preconditions.checkState(
+          anonymousArrayExpr.exprs() != null && anonymousArrayExpr.exprs().size() > 0,
+          EMPTY_EXPRS_MESSAGE);
+      Reference ref = anonymousArrayExpr.exprs().get(0).type().reference();
+      Preconditions.checkNotNull(ref, "Annotations must be an Object type");
```

Comment:
I think we should get the type info from `TypeNode` field and use it as the source of truth for all checks.